Add Prepare stage for inputs

Aflynn50 opened the pull request:
This PR contains the Prepare stage for input parts. In the prepare stage the user passes instantiations of every Go struct that is mentioned in an input or output expression. These objects are used only for information about their type, the actual values inside the structs are irrelevant. 

The `prepareInput` function checks that the structs and tags mentioned in IO expressions in the query correspond to the types of the objects passed as arguments to `Prepare`.

This PR is an updated version of #24 which was a draft PR. However since the name of the branch needed changing this PR has been created.

TheSignPainter98 commented on internal/prepare/prepare.go:
```diff
@@ -0,0 +1,76 @@
+type typeNameToInfo map[string]*typeinfo.Info
+
+// prepareInput checks that the input expression corrosponds to a Go struct
+// passed to Prepare.
```
s/corrosponds/corresponds/

Aflynn50 replied:
Done.

TheSignPainter98 commented on internal/prepare/prepare_test.go:
```diff
@@ -0,0 +1,67 @@
+
+func (s *ParserSuite) TestMissingTagInput(c *C) {
+	sql := "SELECT street FROM t WHERE x = $Address.number"
+	parser := parse.NewParser()
```
This isn't within the scope of this PR, but `parse.NewParser()` feels slightly off to me, perhaps renaming the `parse` module to `syntax` would be clearer (this would become `syntax.NewParser()`). Being a noun, this module name would match the existing `typeinfo` module, but be at odds with the existing `prepare` module (verb).

Aflynn50 replied:
It does feel off, I'm currently reading my way through the google style guide and it suggests using something like `parse.New` to avoid repetition. At some point we're going to do a big PR that makes it conform to the  The `typeinfo` module was originally `reflect` which made more sense, but clashed with the standard library.

TheSignPainter98 replied:
Ah right, I understand.

I actually really like the idea of using `parse.New` (or something similar), it feels very nice and concise.

TheSignPainter98 commented on internal/prepare/prepare.go:
```diff
@@ -0,0 +1,76 @@
+// Prepare takes a parsed expression and all the go objects mentioned in it.
+// The IO parts of the statement are checked for validity against the Go objects
+// and expanded if necessary.
+func Prepare(pe *parse.ParsedExpr, args ...any) (expr *PreparedExpr, err error) {
```
The capitalisation on 'go' is inconsistent here :P 

Also this comment doesn't really describe _why_ `Prepare` would be used, it seems to be a re-phrasing of what the code is doing. Following a [previous conversation] on #24, perhaps something like this would be good?
```go
// Prepare prepares a parsed expression for execution in SQL.
```
[previous conversation]: https://github.com/canonical/sqlair/pull/24#discussion_r1048450990

Aflynn50 replied:
Fixed, and comment added.

TheSignPainter98 commented on internal/prepare/prepare.go:
```diff
@@ -0,0 +1,76 @@
+		sql.WriteString(p.ToSQL())
+	}
+
+	return &PreparedExpr{ParsedExpr: pe, SQL: strings.TrimSpace(sql.String())}, nil
```
Is this `TrimSpace` necessary? If it is, it sounds like `sql.String()` is putting in too many spaces for some reason,  so should probably be fixed elsewhere.

Aflynn50 replied:
True, this can actually be removed. Previously we had been enforcing correct spacing around queries, but now its been changed to accept whatever was there before. The `TrimSpace` took care of spaces round IO expressions at the start of end of the string.

solmyrcan commented on internal/prepare/prepare.go:
```diff
@@ -0,0 +1,76 @@
+// prepareInput checks that the input expression corresponds to a Go struct
+// passed to Prepare.
+func prepareInput(ti typeNameToInfo, p *parse.InputPart) error {
+	if inf, ok := ti[p.Source.Prefix]; ok {
```
I read about this recently - dealing with error cases first and return, and then leave the "success" case and its treatment after those checks are done, rather than success being within an if block. Perhaps this approach could be adopted here? :) 

```
if inf, ok := ti[p.Source.Prefix]; !ok {
    return fmt.Errorf("unknown type: %s", p.Source.Prefix)
}
if _, ok := inf.TagToField[p.Source.Name]; !ok {
    return fmt.Errorf("there is no tag with name %s in %s", p.Source.Name, inf.Type.Name())
}
// success
return nil
```

Aflynn50 replied:
Very true! I also read this recently - I've changed it.

solmyrcan commented on internal/prepare/prepare.go:
```diff
@@ -0,0 +1,76 @@
+
+	// Generate and save reflection info.
+	for _, arg := range args {
+		i, err := typeinfo.TypeInfo(arg)
```
Minor comment, but would it make sense to use something like "info" rather than "i" here, as I guess traditionally "i" is used as an index variable name.

Aflynn50 replied:
Maybe, but I think since its scope is so small its probably ok there, and then its consistant with `ti` which is Type Info.

Aflynn50 replied:
I've changed my mind, I think `inf` is actually better lol.

solmyrcan commented on internal/prepare/prepare.go:
```diff
@@ -0,0 +1,76 @@
+			sql.WriteString(p.ToSQL())
+			continue
+		}
+		p := part.(*parse.BypassPart)
```
Would we need to check the outcome of the type assertion here as well? If I understand it correctly, the type could only be one of `*parse.InputPart`, `*parse.OutputPart`, and `*parse.BypassPart`, but from an error handling point of view, would `Prepare` need to handle any case in which the input is actually not one of these, for reliability.

Aflynn50 replied:
True, it does seem unlikely that another would be added, but since in Go its pretty easy to accidentally make something part of an interface its probably worth adding this anyway.

solmyrcan commented on internal/prepare/prepare_test.go:
```diff
@@ -0,0 +1,67 @@
+	parser := parse.NewParser()
+	parsedExpr, err := parser.Parse(sql)
+	_, err = prepare.Prepare(parsedExpr, Person{ID: 1})
+	c.Assert(err, ErrorMatches, "cannot prepare expression: unknown type: Address")
```
I may not have understood the test here, so please ignore if not relevant :) In line, 57, we deliberately pass a `Person` to `Prepare()`, so should it complain about the `sql` statement having an Address or about the second argument being a `Person`? Does this test have the expectation that `Prepare()` must complain when it receives an expression that does not contain the provided type (`Person`, in this case). I guess the error is produced by `prepareInput()`. Could that method equivalently complain about the provided argument type itself? or do we always blame the sql expression :P

Aflynn50 replied:
Prepare basically takes all the type arguments it gets into a bag and looks in the bag for the ones it finds in the query. So if extra arguments are given then they are ignored.

Admittedly this behaviour isn't really documented, so I will add that!

Aflynn50 committed:
Add prepare functions for inputs

Change comments and remove TrimSpace

Minor refactor and some additional comments

Add back prepare

Refactor prepareInput

Change error wording

Tidy up comments

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,73 @@
+// PreparedExpr contains an SQL expression that is ready for execution.
+type PreparedExpr struct {
+	ParsedExpr *ParsedExpr
+	SQL        string
```
I would make those details private, as one of them is the input to the query, and the other one is the actual "prepared" expression. If we expose it, we'll lose the opportunity to do anything else with this type without breaking its promises.

Aflynn50 replied:
I've changed `ParsedExpr` to `inputs` and made it private. I've left SQL as public because we use it in testing and it makes sense that the `PreparedExpr` should have the prepared SQL.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,73 @@
+func prepareInput(ti typeNameToInfo, p *inputPart) error {
+	inf, ok := ti[p.source.prefix]
+	if !ok {
+		return fmt.Errorf(`unknown type: "%s"`, p.source.prefix)
```
Instead of ˋ"%s"ˋ you can use %q. The latter will handle escaping.

That said, we should not need the quoting here as these names shoud be validated elsewhere, and perhaps we could make the debugging easier with something along the lines of:
```
cannot prepare expression: type Fooz unknown, have: Foo, Bar, Baz
```

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	if _, ok = inf.tagToField[p.source.name]; !ok {
+		return fmt.Errorf(`no tag with name "%s" in "%s"`,
```
Similarly, maybe:
```
cannot prepare expression: type Fooz has no "field_name" db tag
```
The field name here needs to be quoted because being a word it could be mixed with the message and misinterpreted.

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,73 @@
+
+	// Generate and save reflection info.
+	for _, arg := range args {
+		inf, err := typeInfo(arg)
```
Can we please use "info" instead of "inf" for the shorthand? I read it as infinite every time. :)

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,73 @@
+	var sql bytes.Buffer
+	// Check and expand each query part.
+	for _, part := range pe.queryParts {
+		if p, ok := part.(*inputPart); ok {
```
This would look a bit nicer with a type switch statement. It also needs a default clause so we get an "internal error" message instead of a panic if we ever do something silly.

Aflynn50 replied:
Changed.

niemeyer reviewed (changes requested):
Probably the only pass. Looks very nice, thanks!

Aflynn50 committed:
Change: error messages, queryPart interface, inf->info

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,86 @@
+		keys[i] = k
+		i++
+	}
+	return keys
```
Are tests passing consistently?  Iterating over the map is unstable, so this needs a `sort.Strings` call before return.

Aflynn50 replied:
I clearly hadn't run it enough times! Changed.

niemeyer commented on internal/expr/prepare.go:
```diff
@@ -0,0 +1,86 @@
+		case *bypassPart:
+			sql.WriteString(p.chunk)
+		default:
+			return nil, fmt.Errorf("internal error")
```
This should say what the internal error is: ("internal error: unknown query part type: %T", part)

Aflynn50 replied:
Changed.

niemeyer reviewed (approved):
A couple of trivials and LGTM, thanks!

Aflynn50 committed:
Sort keys and add more explanative error
